plugins/kubernetes/test/controllers/kubernetes_dashboard_controller_test.rb

```diff
@@ -36,30 +44,112 @@
-      data.each { |item| delete_ids_from item }
+  def pod_list(cluster_deploy_group)
+    empty_list_template.tap do |list|
+      release.release_docs.each do |release_doc|
+        list.items << pod_item(cluster_deploy_group, release_doc) if release_doc.deploy_group == cluster_deploy_group.deploy_group
+      end
     end
   end
+
+  def pod_item(cluster_deploy_group, release_doc)
+    pod_list_item(release_doc.kubernetes_release.project, release_doc.kubernetes_release, release_doc.kubernetes_role, cluster_deploy_group)
+  end
+
+  def empty_list_template
+    RecursiveOpenStruct.new({
+      'kind': 'PodList',
+      'apiVersion': 'v1',
+      'items': []
+    })
+  end
+
+  def pod_list_item(project, release, role, cluster_deploy_group)
+    {
+      'metadata': {
+        'name': 'pod-name',
+        'namespace': cluster_deploy_group.namespace,
+        'labels': {
+          'project_id': project.id,
+          'release_id': release.id,
+          'deploy_group_id': cluster_deploy_group.deploy_group.id,
+          'role_id': role.id
+        }
+      },
+      'status': {
+        'phase': 'Running',
+        'conditions': [
+          {
+            'type': 'Ready',
+            'status': 'True'
+          }
+        ]
+      }
+    }
+  end
+
+  def expected_response
+    [].tap { |response|
+      release.release_docs.each { |release_doc|
+        add_role(response, release_doc).tap { |role|
+          add_deploy_group(role, release_doc).tap { |deploy_group|
+            add_release(deploy_group, release_doc)
+          }
+        }
+      }
+    }
+  end
+
+  def add_role(response, release_doc)
+    response << role_template(release_doc.kubernetes_role) unless response.any? { |role|
+      role.id == release_doc.kubernetes_role.id
+    }
+
+    response.find { |role| role.id == release_doc.kubernetes_role.id }
+  end
+
+  def add_deploy_group(role, release_doc)
+    role.deploy_groups << deploy_group_template(release_doc.deploy_group) unless role.deploy_groups.any? { |deploy_group|
+      deploy_group.id == release_doc.deploy_group.id
+    }
+
+    role.deploy_groups.find { |deploy_group| deploy_group.id == release_doc.deploy_group.id }
+  end
+
+  def add_release(deploy_group, release_doc)
+    deploy_group.releases << release_template(release, release_doc) unless deploy_group.releases.any? { |release|
+      release.id == release_doc.kubernetes_release.id
+    }
+  end
+
+  def role_template(role)
+    RecursiveOpenStruct.new({
+      'id': role.id,
+      'name': role.name,
+      'deploy_groups': []
+    })
+  end
+
+  def deploy_group_template(deploy_group)
+    RecursiveOpenStruct.new({
+      'id': deploy_group.id,
+      'name': deploy_group.name,
+      'releases': []
+    })
+  end
+
+  def release_template(release, release_doc)
+    RecursiveOpenStruct.new({
+      'id': release.id,
+      'build': release.build.label,
+      'target_replicas': release_doc.replica_target,
+      'live_replicas': 1
+    })
+  end
```

Comment:
I used a JSON template file for 2 reasons:
- It's easy to see the actual structure of the data, which could be useful for someone new starting work on this area of the code.
- You could have an error in your code which generates the requests and expected responses, causing the tests to return false positives. I wouldn't be sure if everything works fine just by looking at the test code.

It's all a matter of preference though :)